### DISCUSSION TITLE
I need help to decide on your API vs. self-hosting and to scale the system

### DISCUSSION THREAD

**Discussion by irmlerc** [open-meteo/open-meteo#519] [Q&A]
Hi,

I have the situation that I'm developing a solution, which I use just personally for myself. But meanwhile, I think more and more that I could maybe also make a product out of it a bit later, which would turn it to commercial then.
In this phase I want to get an idea how I can scale.

My problem: If I should turn to commercial, it is impossible to predict the number of customers and so the number of requests. It could be that there will be zero customers, which would mean that I would have to cancel the commercial project and go back to private, as the original reason for the development is a private reason and I want to use it personally myself. I hope a "downgrading" like that is possible?
But it could also be that I would need to cascade even the API Professional 5-20 times, or even more. Even 60-80 times would be possible in a "very best case" scenario. I also can't tell how fast it grows then. I guess, it is likely to stay below 1Mio requests in the beginning, but what if there would come 1Mio more requests with a sudden? 

In general, I think, it would make sense to start the commercial project then with one of your hostes API's.

Now, to find out how I can scale, I have the following questions:

1) Question to your hosted API:
- How many requests can the "API Standard" handle per minute/hour/day? (to find out how long it takes to process my amount of data)
- How many requests can the "API Professional" handle per minute/hour/day? 
- How many times can the "API Professional" be cascaded maximum?
- Does the Standard and Professional API offer additional data that is not included in the free API? Something like fog-possibility, mist-possibility and cloud-inversion?

2) Questions to self-hosting:
- Is there also a license necessary (software + weather data) when I use my own server?
- How much traffic will I produce when I update the weather all 3 hours (25 weather phenomenons of type "hourly")?
- These weather data is not needed just for 1 single location, it could be that I would need them for thousands, ten thousands, or hundred thousands of locations 
- How much disk space will I need for that use case?
- What are the system requirements (CPU, memory) for that use case? (does it run on a Virtual Machine with SSD, where I can increase CPU/memory whenever I need?)
- Where can I find an install guide for self-hosting on Debian? Or is Ubuntu recommended?
- Where can I find guides for updating when self-hosted?
- How many requests can a strong self-hosted API handle per minute/hour/day? (please example with CPU/RAM and how much it can do)

Sorry for the many questions, but I really need to know that to get an idea of how I should start.
Thanks a lot!

**Comment by patrick-zippenfenig** [comment]
Hi, thanks for asking!

1. API subscription

There are no fixed minutely/daily/hourly limits for the API subscriptions. However, the concurrency per API subscriber might get limited to 4 concurrent API calls. In this case the API latency would gradually increase without disrupting any customer application.

You can purchase multiple API subscriptions to increase the number of calls. However, you can also contact me directly and get a volume discount or even dedicated API nodes.

No, there is no additional data for paying customers. Currently one exception is that the regular forecast API offer less past data while using the `past_days` parameter. The reason is that the free API nodes have less disk space.

2. Self hosted

No license is required. For certain provider you will need an API key to retrieve data (e.g. ECMWF Copernicus). 

The amount of traffic is hard to estimate and depends on which weather models and weather variables are used. For all forecast models the API roughly downloads 1 TB per day. It is not possible to limit the area, because weather models are global or have fixed regions. However if you only use temperature on 2 meters above ground from a small selection of models, it can be as low as 1 GB per day.

Disk space depends as well on configuration. I typically use API nodes with 1 TB to keep data for the past 3 months.

I would recommend at least 4 CPU cores with 16 GB memory. The main limitation is downloading and processing data. The actual "API" works with 1GB memory. In the future, the converted Open-Meteo database might be available as open-data directly. However, there is no ETA yet.

Debian is not supported. The main limitation is the `eccodes` library that needs to be up to date. I only bundle it with ubuntu. https://github.com/patrick-zippenfenig/ecCodes-ubuntu

Updates can be performed with APT or Docker. Please note, that there will be breaking changes in the file and database layout. Make sure to read the latest release notes.

API performance depends on your use-case and specific hardware configuration. Fast disks and plenty of memory improve performance, but you might also get CPU bottlenecked. You will have to do benchmarks yourself and find the optimal hardware configuration.


Best, Patrick

**Reply by irmlerc** [reply]
Hi Patrick,

Thank you for your detailed answer!

>the concurrency per API subscriber might get limited to 4 concurrent API calls

There will be no concurrent API calls from my side, they will all come sequentially (when one is received, there will come the next one). It is more that I want to find out how long it takes to receive, let's say 100.000 calls one after the other with 20 hourly data fields. Just to find out if I'm able to receive all the data I need within my time window :)
Could you please tell me here any times that I can estimate?

>You can purchase multiple API subscriptions to increase the number of calls.

### So, there is no limit to cascade subscriptions from your side, it is just limited technically, by the time how long it takes to receive the data, right? I really need to know how many calls I can maximum receive within an hour, or how long 10.000 or 100.000 calls (sequentially) in one go will take, for instance when we would define 20 hourly data fields.

>No, there is no additional data for paying customers.

### One question to that:
If I need the "forecast" api endpoint for ~20 data fields, but I would also need e.g. the wave heights from the marine forecast endpoint (because it is not included in the "normal" forecast): Is that marine forecast endpoint included in my API license for the normal forecast? Or is there an extra license necessary?


>No license is required. For certain provider you will need an API key to retrieve data (e.g. ECMWF Copernicus).

### And which costs gets occurred by that API key?
I'm pretty sure I will start with your paid API. The question is just, how many calls I can get over, as I asked above. To find out when the point comes, where your offered paid API gets "too small" for me.


>However if you only use temperature on 2 meters above ground from a small selection of models, it can be as low as 1 GB per day.

### OK, this means we are talking around ~1GB per hourly data field, right? So, when I use 20 data fields, for instance, we are talking about 20GB per day, if I update all 3 hours?
And that just if I receive data up to multiple days, I guess. So, when I just receive the forecasts for the next 48 hours (=what I need), it will get even reduced massively?
In that case I could keep the traffic extremely down, when I just update twice a day and there just for 48 hour forecasts. In that case it would be possible to reduce it down to just 2-3GB a day for 20 data fields. Do I see that right?

>Disk space depends as well on configuration. I typically use API nodes with 1 TB to keep data for the past 3 months.

### I don't need old data stored on my disk, I'm just interested in the forecasts of the upcoming 48 hours. Actually, I could delete all the data that is in the past. Then I guess, we are just talking about something like 20-30GB? 

If you please could answer these points, as well.
As soon as it gets concrete to turn it commercial, I will reach out to you personally over your website.

And just in case the commercial project would be a flop, I could switch back to a free personal license? Again, my main aim is to use it myself. Turning it commercial is just an idea, but which gets more and more clearer.

Thank you!

**Reply by patrick-zippenfenig** [reply]
> Could you please tell me here any times that I can estimate?

For 20 hourly fields I would estimate the performance to around 200 locations per second. 10.000 locations roughly 1 minute. If you are trying to generate a map, you might also consider downloading the original GRIB files from NOAA or DWD.

> Is that marine forecast endpoint included in my API license for the normal forecast? Or is there an extra license necessary?

Only weather forecasts from JMA and ECMWF historical archives require additional subscriptions. Other data can be accessed for free.

> OK, this means we are talking around ~1GB per hourly data field, right? So, when I use 20 data fields, for instance, we are talking about 20GB per day, if I update all 3 hours? 

Only 20 GB traffic seems a bit low. It is probably a bit more. Again, it is hard to estimate, because I only download all data on my nodes. Yes, it includes all updates.

> And that just if I receive data up to multiple days, I guess. So, when I just receive the forecasts for the next 48 hours (=what I need), it will get even reduced massively?

No, if you run your own API nodes, you will have to download the entire global model and forecast range. Theoretically, it would be possible to only download 48 hours, but there is no support for that in my code. Most use-cases require more than 48 hours of forecast.

> Actually, I could delete all the data that is in the past. Then I guess, we are just talking about something like 20-30GB?

Possible. However, data is organised into chunks of ~2 weeks data. Therefore, you are forced to keep ~4 weeks of data.

> And just in case the commercial project would be a flop, I could switch back to a free personal license?

Sure, you can cancel the subscription any time. Many users just get the API subscription for one month to fetch historical data and cancel afterwards.

**Comment by meb427** [comment]
For 2

Did you check the docs? https://github.com/open-meteo/open-meteo/blob/main/docs/getting-started.md
It's all there how to host the app and there is no licence as you can see.
Disk space requirement if you pull all the models witthout historical data will be 500GB-600GB. No idea about the ingress bw need. But you need a  solid disk to process all data. You can start locally on your laptop if you have 1TO free.
You can start with 1 VM or host that have the API and data conversion/retrieval, but as Patrick would point and also offered the feature. You can have later conversion heavy on different server or even 2 different that will sync once updated with main server.
Didn't do a benchmark but the API is quite fast and performance will depend heavily on your specs CPU/Disk ( less RAM), also in case of heavy load you can use a load balancer in front of it.

**Reply by irmlerc** [reply]
Hi @meb427,
Thank you for your kind reply!
You are right, it seems that it indeed makes sense to try it out on a fresh VM on my local computer, just to find out how much space is used. Thank you for the tip!
